aiida_common_workflows/workflows/eos.py

```diff
@@ -0,0 +1,158 @@
+
+    @classmethod
+    def define(cls, spec):
+        # yapf: disable
+        super().define(spec)
+        spec.input('structure', valid_type=orm.StructureData, help='The structure at equilibrium volume.')
+        spec.input('scale_factors', valid_type=orm.List, required=False, validator=validate_scale_factors,
+            help='The list of scale factors at which the volume and total energy of the structure should be computed.')
+        spec.input('scale_count', valid_type=orm.Int, default=lambda: orm.Int(7), validator=validate_scale_count,
+            help='The number of points to compute for the equation of state.')
+        spec.input('scale_increment', valid_type=orm.Float, default=lambda: orm.Float(0.02),
+            validator=validate_scale_increment,
+            help='The relative difference between consecutive scaling factors.')
+        spec.input('generator_inputs', valid_type=orm.Dict, validator=validate_generator_inputs,
+            help='The inputs that will be passed to the inputs generator of the specified `sub_process`. It should '
+            'include all the inputs, excluding the `structure` and `previous_workchain`.')
+        spec.input_namespace('sub_process', dynamic=True, populate_defaults=False)
+        spec.input('sub_process_class', non_db=True, validator=validate_sub_process_class)
+        spec.inputs.validator = validate_inputs
+        spec.outline(
+            cls.run_init,
+            cls.run_eos,
+            cls.inspect_eos,
+        )
+        spec.output_namespace('structures', valid_type=orm.StructureData,
+            help='The relaxed structures at each scaling factor.')
+        spec.output_namespace('total_energies', valid_type=orm.Float,
+            help='The computed total energy of the relaxed structure at each scaling factor.')
+        spec.exit_code(400, 'ERROR_SUB_PROCESS_FAILED',
+            message='At least one of the `{cls}` sub processes did not finish successfully.')
+
+    def get_scale_factors(self):
+        """Return the list of scale factors."""
+        if 'scale_factors' in self.inputs:
+            return self.inputs.scale_factors
+
+        count = self.inputs.scale_count.value
+        increment = self.inputs.scale_increment.value
+        return [orm.Float(1 + i * increment - (count - 1) * increment / 2) for i in range(count)]
+
+    def get_sub_workchain_builder(self, scale_factor, previous_workchain=None):
+        """Return the builder for the relax workchain."""
+        structure = scale_structure(self.inputs.structure, scale_factor)
+        process_class = WorkflowFactory(self.inputs.sub_process_class)
+        generator_inputs = self.inputs.generator_inputs.get_dict()
+
+        builder = process_class.get_inputs_generator().get_builder(
+            structure,
+            calc_engines=generator_inputs['calc_engines'],
+            protocol=generator_inputs['protocol'],
+            relaxation_type=RelaxType(generator_inputs['relaxation_type']),
```

Comment:
It is a shame we can not simply use `**generator_inputs`. The reason is the fact we need to call `RelaxType`. How do we remember to change this call whenever we change the signature of `get_builder`?

Reply:
> It is a shame we can not simply use `**generator_inputs`. The reason is the fact we need to call `RelaxType`.

Very true, and I think this is again a manifestation of the limitation of the concept of separation of input generation and workchain logic implementation. The only thing I can think of, for now, is to change the `generate_inputs` into a `non_db` namespace and let us just pass whatever value we would pass directly onto 

> How do we remember to change this call whenever we change the signature of `get_builder`?

Normally this is done through tests. I haven't added any here because not trivial to write them

Reply:
Ok, I'll write an issue saying we have to write tests to remind us this and for the moment we leave this like it is. We should definitely have a call one day and discuss in depth the limits and advantages of our approach and see if there are alternatives.